allennlp/training/trainer.py

```diff
@@ -980,11 +980,34 @@ def from_params(cls,
         grad_norm = params.pop_float("grad_norm", None)
         grad_clipping = params.pop_float("grad_clipping", None)
         lr_scheduler_params = params.pop("learning_rate_scheduler", None)
+        fine_tuning = params.pop("fine-tuning", None)
+        optimizer_params = params.pop("optimizer")
+        if fine_tuning == "discriminative" or \
```

Comment:
It's not great to have all of this logic inside of a `from_params` method.  I'm not sure that there's a great place for it, though, as you're modifying other parameters based on a flag.  At the very least, you could pull this logic out into a `create_parameter_groups()` method, or similar, and then you could test that method in a simple unit test.

Reply:
see above, I would rather this logic live inside the `Optimizer` constructor (modulo the fact that there is no such constructor, maybe we need a Wrapper class like we do for `Seq2Seq` etc...)

Reply:
also, one stylistic comment, we generally try to avoid using `\` for line continuations, preferring to use parenthesis for expressions that need to span multiple lines

Reply:
So, for this PR and @sebastianruder's contribution here, should we pull this out into a function and try to fix the broader issue later?  I don't think Sebastian can or wants to solve the mess we've made of `Trainer.from_params`.

Reply:
yeah, that shouldn't be part of this PR, I think pulling it out into a function (and adding some clarifying comments) is good for now. 

fixing the from_params mess is a broader issue

Reply:
I think the best way to fix this (not for this PR) would be to build the optimizer without modifying it's parameters, and then modify the `parameter_groups` of the optimizer directly inside the learning rate scheduler, as the optimizer is passed to the LR scheduler as an argument. This would isolate the implementation to where it is actually used (i.e in the special scheduler) and also solve the "you can't use this unless you build from params" problem, which is something we must avoid.

Reply:
Is "discriminative fine tuning" just a property of the learning rate scheduler, or is it a property of something else?  If it really is just a property of the scheduler, then I definitely agree with @DeNeutoy, and I'd say that this change _does_ belong in this PR.

Reply:
The only issue is if you can't update the parameters groups of an optimizer after-the-fact.  I don't know enough about pytorch optimizers to know if this would break things.

Reply:
I agree with @DeNeutoy. I've moved the discriminative fine-tuning now from `trainer.py` to the learning rate scheduler (see https://github.com/allenai/allennlp/pull/1636/commits/a81040279d701163d50f8798631c1faec44f7c16). This is a lot cleaner and a lot more intuitive.